### DISCUSSION TITLE
AnalogContinuous PlatformIO and ArduinoIDE(1.8.19)

### DISCUSSION THREAD

**Discussion by ceewanna** [espressif/arduino-esp32#10504] [General]
Based on this [https://docs.espressif.com/projects/arduino-esp32/en/latest/installing.html](url)
I was running on PlatformIO and was trying to test ADC's analogContinuous on [https://docs.espressif.com/projects/arduino-esp32/en/latest/api/adc.html](url)

1. In PlatformIO it couldn't find necessary includes.
2. So I moved to ArduinoIDE (1.8.19). While installing esp32 board manager, I found that 3.0.6 couldn't get through installation and had this error below while 3.0.5 was installed successfully. 
![image](https://github.com/user-attachments/assets/64e453a6-0d49-4d03-8d82-59320703a916)

3. AnalogContinuous was compiled successfully in ArduinoIDE(1.8.19) and recognized right away without any extra includes but not  in PlatformIO. 

My PlatformIO platformio.ini had this 
```
[env:Espressif32]

platform = espressif32
framework = arduino

[env:Espressif32dev]
platform = https://github.com/platformio/platform-espressif32.git
framework = arduino


[env:Espressif32latest]
platform = https://github.com/platformio/platform-espressif32.git
framework = arduino
platform_packages =
    framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32#master

[env:c3]
extends = env:Espressif32latest
board = esp32-c3-devkitm-1  ; c3 super mini
upload_speed = 921000
monitor_speed = 115200
```

None of them could recognize analogContinuous and env:Espressif32latest has error when added to platformio.ini
env:Espressif32 and env:Espressif32dev have no errors but failed to recognize analogContinuous.

```
Resolving c3 dependencies...
Already up-to-date.
Updating metadata for the vscode IDE...
UserSideException: Processing c3 (board: esp32-c3-devkitm-1; platform: https://github.com/platformio/platform-espressif32.git; framework: arduino)
--------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32-c3-devkitm-1.html
PLATFORM: Espressif 32 (6.9.0+sha.61e9531) > Espressif ESP32-C3-DevKitM-1
HARDWARE: ESP32C3 160MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-builtin, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-arduinoespressif32 @ 3.0.5+sha.4285912 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - toolchain-riscv32-esp @ 8.4.0+2021r2-patch5
KeyError: 'framework-arduinoespressif32-libs':
  File "/home/cw/.platformio/penv/lib/python3.11/site-packages/platformio/builder/main.py", line 173:
    env.SConscript("$BUILD_SCRIPT")
  File "/home/cw/.platformio/packages/tool-scons/scons-local-4.8.1/SCons/Script/SConscript.py", line 620:
    return _SConscript(self.fs, *files, **subst_kw)
  File "/home/cw/.platformio/packages/tool-scons/scons-local-4.8.1/SCons/Script/SConscript.py", line 280:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "/home/cw/.platformio/platforms/espressif32@src-5f117260f75b328038ec9d3fd0e14a68/builder/main.py", line 368:
    target_elf = env.BuildProgram()
  File "/home/cw/.platformio/packages/tool-scons/scons-local-4.8.1/SCons/Util/envs.py", line 252:
    return self.method(*nargs, **kwargs)
  File "/home/cw/.platformio/penv/lib/python3.11/site-packages/platformio/builder/tools/piobuild.py", line 61:
    env.ProcessProgramDeps()
  File "/home/cw/.platformio/packages/tool-scons/scons-local-4.8.1/SCons/Util/envs.py", line 252:
    return self.method(*nargs, **kwargs)
  File "/home/cw/.platformio/penv/lib/python3.11/site-packages/platformio/builder/tools/piobuild.py", line 121:
    env.BuildFrameworks(env.get("PIOFRAMEWORK"))
  File "/home/cw/.platformio/packages/tool-scons/scons-local-4.8.1/SCons/Util/envs.py", line 252:
    return self.method(*nargs, **kwargs)
  File "/home/cw/.platformio/penv/lib/python3.11/site-packages/platformio/builder/tools/piobuild.py", line 347:
    SConscript(env.GetFrameworkScript(name), exports="env")
  File "/home/cw/.platformio/packages/tool-scons/scons-local-4.8.1/SCons/Script/SConscript.py", line 684:
    return method(*args, **kw)
  File "/home/cw/.platformio/packages/tool-scons/scons-local-4.8.1/SCons/Script/SConscript.py", line 620:
    return _SConscript(self.fs, *files, **subst_kw)
  File "/home/cw/.platformio/packages/tool-scons/scons-local-4.8.1/SCons/Script/SConscript.py", line 280:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "/home/cw/.platformio/platforms/espressif32@src-5f117260f75b328038ec9d3fd0e14a68/builder/frameworks/arduino.py", line 36:
    SConscript(
  File "/home/cw/.platformio/packages/tool-scons/scons-local-4.8.1/SCons/Script/SConscript.py", line 684:
    return method(*args, **kw)
  File "/home/cw/.platformio/packages/tool-scons/scons-local-4.8.1/SCons/Script/SConscript.py", line 620:
    return _SConscript(self.fs, *files, **subst_kw)
  File "/home/cw/.platformio/packages/tool-scons/scons-local-4.8.1/SCons/Script/SConscript.py", line 280:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "/home/cw/.platformio/packages/framework-arduinoespressif32@src-e36c2260a49b9bf6768237494cebb35d/tools/platformio-build.py", line 38:
    FRAMEWORK_LIBS_DIR = platform.get_package_dir("framework-arduinoespressif32-libs")
  File "/home/cw/.platformio/penv/lib/python3.11/site-packages/platformio/platform/_packages.py", line 32:
    pkg = self.get_package(name)
  File "/home/cw/.platformio/penv/lib/python3.11/site-packages/platformio/platform/_packages.py", line 29:
    return self.pm.get_package(spec or self.get_package_spec(name))
  File "/home/cw/.platformio/penv/lib/python3.11/site-packages/platformio/platform/_packages.py", line 21:
    owner=self.packages[name].get("owner"),
========================== [FAILED] Took 0.31 seconds ==========================

```

What is missing in PlatformIO? or do I have to move everything back to ArduinoIDE?

**Comment by sivar2311** [comment]
AnalogContinuous is a function from Espressif32 Arduino Core 3.x.
You missed that there is no official Espressif32 Arduino 3.x available on PlatformIO. See https://github.com/platformio/platform-espressif32/issues/1225
The latest available Espressif32 Arduino Core on PlatformIO is 2.0.17 by using `platform = espressif32 @ 6.9.0`

But the [AnalogContinuous example](https://docs.espressif.com/projects/arduino-esp32/en/latest/api/adc.html#example-applications) works great with the (unofficial) Arduino Core 3.0.6 provided by [pioarduino](https://github.com/pioarduino/platform-espressif32):
`platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.06/platform-espressif32.zip`

List of [available versions](https://gist.github.com/sivar2311/5125011eaaf859eff90c4b502e1408b1#file-platform-espressif32-versions-md)

**Reply by ceewanna** [reply]
@sivar2311 Thank you so much for showing me pioarduino. It's the first time I have seen it. I will give it a try.

I was struggling on using arduino ide/cli from vscode/vscodium after facing the problem. It's really a painful process trying to include folders after folders in the json file. 

Strange but true, Espressif still recommends both arduino ide and platformio in their official page even though the leg of platformio is already outdated.

**Comment by Jason2866** [comment]
Latest official espressif32 Arduino core 3.0.7 is available via pioarduino fork
```
platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.07/platform-espressif32.zip
```

**Reply by ceewanna** [reply]
Thank you. By the way will this work for ESP8684 or ESP32C2 too?

**Reply by Jason2866** [reply]
No, since it is 100% espressif32 Arduino. There are no compiled libs for the C2

**Reply by ceewanna** [reply]
I used to come across somewhere and made use of the setup below from your work. With the problem I faced above, I started to wonder what direction I should take given that I will working on ESP32, ESP32S2, ESP32C3 and ESP32C2.
```
[env:esp32c2]
platform = https://github.com/Jason2866/platform-espressif32.git#Arduino/IDF5
framework = arduino
board = esp32-c2-devkitm-1
```

**Reply by sivar2311** [reply]
@ceewanna See https://github.com/espressif/arduino-esp32?tab=readme-ov-file#supported-chips and the Note below

**Reply by Jason2866** [reply]
@ceewanna My fork has major changes in settings and supported IDF components. The source code is same as for official Arduino core. BUT many stuff is removed to keep resource use low for my needs.

**Reply by ceewanna** [reply]
@sivar2311 So it means pioarduino will follow whatever supported by the official espressif arduino esp32?

@Jason2866 Your fork will enable us to make use of both Arduino and ESPIDF?

**Reply by Jason2866** [reply]
Yes, the corresponding IDF version to the Arduino version.
Branch https://github.com/Jason2866/platform-espressif32/tree/Arduino/IDF53  is Arduino core 3.1.0 and IDF 5.3.1 But keep in mind, it is my working branch and there are often updates and it happens it is totally broken for a while. It is always bleeding edge! To be clear it is not made to be used for others. You can, but no support of any kind.

**Reply by Jason2866** [reply]
I will probably provide in future a pioarduino version where the Arduino libs are compiled just in time before the Arduino compile run. When my early version of this new build variant has stabilised and left bugs are removed i will port to pioarduino. When it is finally available the C2 can be used with official Arduino core without doing anything special... And no idk when it is ready. It is done when it is done.

**Reply by ceewanna** [reply]
@Jason2866  That sounds great. By the way, I am now using this for ESP32C3
```
platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.07/platform-espressif32.zip
```
It could compile and build but there are these shown in Problems tab. I tried changing among versions 51.03.05 to 51.03.07 but it seems the problems remain.

```
[10/25/2024, 9:25:20 PM] Unable to resolve configuration with compilerPath "/home/cw/.platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc".  Using "/usr/bin/gcc" instead.
[10/25/2024, 9:25:20 PM] Unable to resolve configuration with compilerPath "/home/cw/.platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc".  Using "/usr/bin/gcc" instead.
[10/25/2024, 9:32:18 PM] For C++ source files, the cppStandard was changed from "gnu++23" to "gnu++17" based on compiler args and querying compilerPath: "/home/cw/.platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc"
```

**Reply by Jason2866** [reply]
@ceewanna Close VSC delete in project folder `.pio`, `.vsc` and when existing `.cache` 
Start VSC Platformio will init the environment and the problems should not be there anymore.
Such stuff are leftovers when switching environments or platforms

**Reply by ceewanna** [reply]
@Jason2866 Thanks for the fix. I guess it's because I have multiple env: settings because my code has to cater for both ESP8266 and ESP32C3. After commenting out all other env's and delete .vscode  and .pio (even .platformio was deleted too), the problems on resolving paths are gone but the error on gnu++23 remains.

**Reply by Jason2866** [reply]
@ceewanna Enable your commented environments. You will have no issue. Is a other (problematic) toolchain in your path?  If you try to change gnu version in an environment setting this can be problematic. There is **no** choice to do this!  Probably one of the env setups are not correct or trying to do unsupported.

**Reply by ceewanna** [reply]
@Jason2866 Do I have to comment out environments not being used and leave only the one used? I never noticed if I have the problems doing multiple environments. My platformio.ini is below. After removing comments from ESP8266 section, the problems show up again. Btw I never made changes to c_cpp_properties.json or any other files apart from my codes.

```
[platformio]
default_envs = c3

;========= ESP8266 ====================
[env:Espressif8266]
platform = espressif8266@3.1.2
framework = arduino

; platform_packages =
;     platformio/framework-arduinoespressif8266 @ https://github.com/esp8266/Arduino.git

; [env:Espressif8266upstream]
; platform = https://github.com/platformio/platform-espressif8266.git
; framework = arduino

[env:d1]
extends = env:Espressif8266
board = d1
upload_speed = 921000
monitor_speed = 115200
; ; upload_flags = -v
; ; board_build.partitions = partitions/partitions.csv
; ; board_build.ldscript = eagle.flash.4m1m.ld

[env:d1_ota]
extends = env:d1
upload_protocol = espota
upload_port = 192.168.1.229
upload_flags = --host_port=17700

[env:nodemcu]
extends = env:ArduinoEspressif8266
board = nodemcu
upload_speed = 921000
monitor_speed = 115200

;========= ESP32 ====================
; [env:Espressif32]
; platform = espressif32
; framework = arduino

; [env:Espressif32dev]
; platform = https://github.com/platformio/platform-espressif32.git
; framework = arduino

[env:Espressif32pioarduino]
platform = https://github.com/pioarduino/platform-espressif32/releases/download/51.03.07/platform-espressif32.zip
; platform = https://github.com/pioarduino/platform-espressif32.git
framework = arduino

[env:c3]
extends = env:Espressif32pioarduino
board = esp32-c3-devkitm-1  
upload_speed = 921000
monitor_speed = 115200
upload_protocol = espota
upload_port = 192.168.1.85

```

**Reply by Jason2866** [reply]
You can have only one platform active for espressif32 and one for espressif8266

**Reply by ceewanna** [reply]
@Jason2866 I have revised it using platformio's interpolation of values or options instead of environments. Now I can work on a chip by selecting default envs. Thanks for your advice.
